### PR TITLE
limit the number of identical sounds playing at once, based on priority ordering

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
Obviously, this PR tries to preserve as much of the original sound code as possible (though I am currently not sure about the order of the `singularity` check vs. the `instances` check).

Another alternative would be to scrap this all and fully adapt Odamex's sound logic. But, I like the way how Woof currently sounds (and I am sure many others do as well) and I am not sure how Odamex sounds in contrast (and if it sounds as generic as Woof or if it sounds a bit "like Odamex", you know? I don't want to copy audible sound characteristics of another source port).

File: src/s_sound.c [hunk 1]
```diff
@@ -242,6 +264,10 @@ static int S_getChannel(const mobj_t *origin, sfxinfo_t *sfxinfo,
    int cnum;
    int lowestpriority = -1; // haleyjd
    int lpcnum = -1;
+   int instances = 0;
+
+   // Sort the sound channels by descending priority levels
+   SDL_qsort(channels, numChannels, sizeof(channel_t), S_CompareChannels);
```
Comment (rfomin) [inline src/s_sound.c:270]: `SDL_qsort` crashes in MSVC build (works in MSYS). A simple `qsort` works for me. Should we use `SDL_qsort`? It's not documented in https://wiki.libsdl.org/

**Review by rfomin — COMMENTED** [review]
I like this a lot, it works well for me, thought I have bad hearing. Great work!

File: src/s_sound.c [hunk 2]
```diff
@@ -252,17 +278,40 @@ static int S_getChannel(const mobj_t *origin, sfxinfo_t *sfxinfo,
...
+       if (channels[cnum].origin == origin &&
+           channels[cnum].loop && loop)
+       {
+         return -1;
+       }
+
+       // Limit the number of identical sounds playing at once
+       if (++instances >= max_instances)
+       {
+         if (priority <= channels[cnum].priority)
```
Comment (rfomin) [inline src/s_sound.c:297]: It seems `if (priority < channels[cnum].priority)` works more correct dsda_parallellimit.wad

**Comment by fabiangreffrath** [comment]
Thank you for your input! It seems that this implementation completely undermines the `singularity` concept, though. Two identical sounds with the same singularity will get caught in the new "not more than three" condition before they even reach the ingularity check.

**Comment by rfomin** [comment]
> It seems that this implementation completely undermines the `singularity` concept, though. Two identical sounds with the same singularity will get caught in the new "not more than three" condition before they even reach the ingularity check.

It sounds the same as before for me, maybe we can remove `singularity` and simplify `s_sound.c` a bit. I think we need a tester with good hearing 😄

**Review by rfomin — APPROVED** [review]
LGTM. Do we need to set the number of channels to 16 by default? 32 works for me, the default in Odamex is also 32.

**Comment by fabiangreffrath** [comment]
Of course, fine with me. Shouldn't be a pain in the ear anymore anyway. 😉

**Comment by rfomin** [comment]
The sound of the plasma gun has changed - try checking with the fire button held down.

**Comment by fabiangreffrath** [comment]
Hm, how does it sound in Odamex?

**Comment by rfomin** [comment]
> Hm, how does it sound in Odamex?

In Odamex it sounds right. If we change `max_instances = 1` this will be fixed, but it seems wrong.

**Comment by fabiangreffrath** [comment]
Does it work with `<=`?

**Comment by rfomin** [comment]
`if (priority <= channels[cnum].priority)` ? No, the same.

**Comment by fabiangreffrath** [comment]
QSort doesn't preserve for order, so the comparison function should return something more meaningful for identical priorities.

**Comment by fabiangreffrath** [comment]
For example, it could `return a->idnum > b->idnum` if `a->priority == b->priority`.

**Comment by fabiangreffrath** [comment]
Is my sorting order wrong?
```diff
--- a/src/s_sound.c
+++ b/src/s_sound.c
@@ -245,7 +245,7 @@ static int S_CompareChannels(const void *arg_a, const void *arg_b)
 
   // Note that a higher priority number means lower priority!
   if (a->priority != b->priority)
-    return a->priority < b->priority;
+    return a->priority > b->priority;
 
   return a->idnum > b->idnum;
 }
```

**Comment by fabiangreffrath** [comment]
Plasma gun sounds normal to me with the latest change. Revenants are a bit loud on the test map, but are at least equally loud in both sides. What do you mean?

**Comment by rfomin** [comment]
> What do you mean?

The latest changes do not affect the plasma gun for me. It's a small thing, but I'm sure people will notice. I recorded two small audio clips.
this PR:

https://user-images.githubusercontent.com/5077629/202349195-c0f6c2a8-3b67-4715-bb08-efc79b77d8c4.mp4

current master:

https://user-images.githubusercontent.com/5077629/202349200-c3d1351c-e83a-4eb5-b1ad-808274690141.mp4

**Comment by fabiangreffrath** [comment]
> The latest changes do not affect the plasma gun for me. It's a small thing, but I'm sure people will notice.

That's obvious, thank you! I think I was once closer to the original. I will have to debug the sorting order a bit further and reintroduce stable sorting. Might take some time, though...

**Comment by rfomin** [comment]
> I will have to debug the sorting order a bit further and reintroduce stable sorting. Might take some time, though...

How about `start_time` (`gametic`) idea from Odamex?

**Comment by fabiangreffrath** [comment]
I am not sure if it helps. We didn't need it before. I am pretty sure I got the `Channels[]` array sorting messed up at one point.

**Comment by rfomin** [comment]
The latest changes fix the plasma gun for me. Sounds good!

**Comment by fabiangreffrath** [comment]
Phew, thanks for confirming! Last step is to get dsda_parallellimit to sound reasonably load and symmetric and then I think we are done here. Couldn't test this one yet.

**Comment by fabiangreffrath** [comment]
Oh, and `<` or `<=` in `priority <= channels[cnum].priority` is still a question. 😁

**Comment by rfomin** [comment]
The dsda_parallellimit sounds reasonable (I'm comparing it to GZDoom) but maybe slightly different from previous commits. It's hard for me to say.

**Comment by rfomin** [comment]
> Oh, and `<` or `<=` in `priority <= channels[cnum].priority` is still a question.

I think `<` sounds better in dsda_parallellimit and closer to previous commits.

**Comment by fabiangreffrath** [comment]
> The dsda_parallellimit sounds reasonable (I'm comparing it to GZDoom) but maybe slightly different from previous commits. It's hard for me to say.

I wouldn't take the previous commits as a reference. The comparison function simply didn't do what it was supposed to do, i.e. sort the sounds by priority. It returned a `boolean`, which is plain wrong for the C stdlib's `qsort()` implementation, which strictly expects `int` values.

**Comment by fabiangreffrath** [comment]
> I think `<` sounds better in dsda_parallellimit and closer to previous commits.

Okay, thanks for this!

**Comment by rfomin** [comment]
Sounds perfect to me! This problem has been bothering me for a long time, even in "casual" WADs (eg annoying lift sound has been fixed in Ancient Aliens). I think it's a big improvement.

**Comment by fabiangreffrath** [comment]
Glad you like it, and thanks for your constructive feedback!

File: src/s_sound.c [hunk 3]
```diff
@@ -228,6 +230,28 @@ static int S_AdjustSoundParams(const mobj_t *listener, const mobj_t *source,
...
+  const channel_t *b = (const channel_t *) arg_b;
+
+  // Note that a higher priority number means lower priority!
+  const int ret = a->priority - b->priority;
+
+  return ret ? ret : (b->idnum - a->idnum);
+}
+
+// How many instances of the same sfx can be playing concurrently
+static const unsigned int max_instances = 3;
```
Comment (kraflab) [inline src/s_sound.c:253]: Shouldn't this be configurable?

Comment (fabiangreffrath) [reply src/s_sound.c:253]: Erm, yes, right. I'll add that tomorrow.

**Pull request merged**